core/internal/cltest/factories.go

```diff
@@ -472,105 +472,57 @@ func MustInsertFatalErrorEthTx(t *testing.T, txStore txmgr.TestEvmTxStore, fromA
 	return etx
 }
 
-func MustAddRandomKeyToKeystore(t testing.TB, ethKeyStore keystore.Eth) (ethkey.KeyV2, common.Address) {
-	t.Helper()
-
-	k := MustGenerateRandomKey(t)
-	MustAddKeyToKeystore(t, k, &FixtureChainID, ethKeyStore)
-
-	return k, k.Address
-}
-
-func MustAddRandomKeyToKeystoreWithChainID(t testing.TB, chainID *big.Int, ethKeyStore keystore.Eth) (ethkey.KeyV2, common.Address) {
-	t.Helper()
-	k := MustGenerateRandomKey(t)
-	MustAddKeyToKeystore(t, k, chainID, ethKeyStore)
+type RandomKey struct {
+	Nonce    int64
+	Disabled bool
 
-	return k, k.Address
+	chainIDs []utils.Big // nil: Fixture, set empty for none
 }
 
-func MustAddKeyToKeystore(t testing.TB, key ethkey.KeyV2, chainID *big.Int, ethKeyStore keystore.Eth) {
-	t.Helper()
-	ethKeyStore.XXXTestingOnlyAdd(key)
-	require.NoError(t, ethKeyStore.Add(key.Address, chainID))
-	require.NoError(t, ethKeyStore.Enable(key.Address, chainID))
-}
-
-// MustInsertRandomKey inserts a randomly generated (not cryptographically
-// secure) key for testing
-// By default it is enabled for the fixture chain
-func MustInsertRandomKey(
-	t testing.TB,
-	keystore keystore.Eth,
-	opts ...interface{},
-) (ethkey.KeyV2, common.Address) {
-	t.Helper()
-
-	chainIDs := []utils.Big{*utils.NewBig(&FixtureChainID)}
-	for _, opt := range opts {
-		switch v := opt.(type) {
-		case utils.Big:
-			chainIDs[0] = v
-		case []utils.Big:
-			chainIDs = v
-		}
+func (r RandomKey) MustInsert(t testing.TB, keystore keystore.Eth) (ethkey.KeyV2, common.Address) {
+	if r.chainIDs == nil {
+		r.chainIDs = []utils.Big{*utils.NewBig(&FixtureChainID)}
 	}
 
 	key := MustGenerateRandomKey(t)
 	keystore.XXXTestingOnlyAdd(key)
 
-	for _, cid := range chainIDs {
-		var nonce int64
-		enabled := true
-		for _, opt := range opts {
-			switch v := opt.(type) {
-			case int:
-				nonce = int64(v)
-			case int64:
-				nonce = v
-			case evmtypes.Nonce:
-				nonce = v.Int64()
-			case bool:
-				enabled = v
-			default:
-				t.Logf("ignoring unknown type in MustInsertRandomKey: %T, note: chain IDs are processed earlier", opt)
-			}
-		}
+	for _, cid := range r.chainIDs {
 		require.NoError(t, keystore.Add(key.Address, cid.ToInt()))
 		require.NoError(t, keystore.Enable(key.Address, cid.ToInt()))
-		err := keystore.Reset(key.Address, cid.ToInt(), nonce)
+		err := keystore.Reset(key.Address, cid.ToInt(), r.Nonce)
 		require.NoError(t, err)
-		if !enabled {
+		if r.Disabled {
 			require.NoError(t, keystore.Disable(key.Address, cid.ToInt()))
 		}
 	}
 
 	return key, key.Address
 }
 
-func MustInsertRandomEnabledKey(
-	t testing.TB,
-	keystore keystore.Eth,
-	opts ...interface{},
-) (ethkey.KeyV2, common.Address) {
-	return MustInsertRandomKey(t, keystore, append(opts, true))
+func (r RandomKey) MustInsertWithState(t testing.TB, keystore keystore.Eth) (ethkey.State, common.Address) {
```

Comment:
Does combining the concerns of inserting a key and then returning the State for that key in this exported API feel right? 

I understand the convenience for consumers in test logic - is it better off as a utility func without `r` as the receiver?

Reply:
:shrug: I wasn't worried about it too much since this is all internal API, and just tried to preserve the existing usages. I open to suggestions, but I think the call site would become verbose if we split it up.

Reply:
Hmm - I've played around with the utility function idea and I think it's worse code quality b/c of the possibilities for expanding the API based on whether the key has a nonce, whether it's disabled, and the chain ids it stores. 

i.e. `MustInsertRandomKeyReturningState` as it is in the PR is the API I was imagining, but doesn't handle the case where you want a specific nonce. It's better for the test writer to specify those fields in the key struct itself rather than expanding the API surface to handle every case, so I think the way you have it in the PR is best.

Reply:
I didn't see your comment before I typed up the response to my initial thought, but I think we're on the same page 👍